### PR TITLE
fix bug of .dockerignore

### LINKED ISSUE #699: buildah bud .dockerignore functionality missing
**Description**

buildah bud needs to check the .dockerignore file in the build directory to exclude files and directories from the build process.

### PR DESCRIPTION
**Pull request by QiWang19**
fix #699 
Exclude(or keep) files when bud with `ADD` or `COPY` in Dockerfile.

Signed-off-by: Qi Wang <qiwan@redhat.com>

### CODE REVIEW

File: imagebuildah/build.go [hunk 1]
```diff
@@ -448,13 +448,53 @@ func (b *Executor) volumeCacheRestore() error {
...
+				for _, p := range excludes {
+					if match, err := filepath.Match(p, filepath.Clean(src)); match {
+						excludeSrc = true
+						break
+					} else if err != nil {
+						return err
+					}
+				}
+			} else {
+				err := filepath.Walk(src, func(path string, info os.FileInfo, err error) error {
```
Comment (kunalkushwaha) [inline imagebuildah/build.go]: With this logic, all files in current folder also will be parsed into this block, and once out of this loop, the value will be just `.` which will be ignored.

maybe you need to append the list of `sources` inside this function.

File: imagebuildah/build.go [hunk 2]
```diff
@@ -448,13 +448,53 @@ func (b *Executor) volumeCacheRestore() error {
...
+					if err != nil {
+						return err
+					}
+					for _, p := range excludes {
+						if match, err := filepath.Match(p, path); match {
```
Comment (giuseppe) [inline imagebuildah/build.go]: should this logic be pushed to `Builder.Add` where we can check each single entry not the src?

**Comment by QiWang19** [comment]
bot, retest this please
:pray:

**Comment by QiWang19** [comment]
bot, retest this please

**Comment by QiWang19** [comment]
@rhatdan PTAL

**Comment by rhatdan** [comment]
LGTM
@giuseppe @nalind @TomSweeneyRedHat @vrothberg PTAL

**Comment by TomSweeneyRedHat** [comment]
@QiWang19 did you test with the the scenario from #699?  I've tried but am having issues.  i think it's unrelated to your changes, but I'm not 100% sure.

**Comment by QiWang19** [comment]
I think it's working

```
[qiwan@qiwan dockerignore-master]$ sudo buildah bud .
STEP 1: FROM quay.io/quay/busybox:latest
Getting image source signatures
Copying blob f0a98344d604 [======================================] 1.4GiB / 1.4GiB
Copying blob 3eecc807f63d [======================================] 1.4GiB / 1.4GiB
Copying blob 48341aa39e69 [======================================] 1.4GiB / 1.4GiB
Copying blob 3eecc807f63d [======================================] 1.4GiB / 1.4GiB
Writing manifest to image destination
Storing signatures
STEP 2: ADD . /tmp/src
STEP 3: RUN ls -l /tmp/src
total 8
-rw-rw-r--    1 root     root            94 Feb  5 14:34 Dockerfile
-rw-rw-r--    1 root     root            12 Feb  5 14:34 source.txt
STEP 4: ENTRYPOINT ["/bin/bash"]
STEP 5: COMMIT containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.mountopt=nodev]@174d00599bba8a22fcb16a29de3dc6d125dec65f6b77d0748a6e9b7ea264a27d
Getting image source signatures
Copying blob f62439784159 [======================================] 512b / 512b
Copying blob de4246eaeb69 [======================================] 2.0KiB / 2.0KiB
Copying blob 35725853fd5a [======================================] 2.7MiB / 2.7MiB
Copying blob de4246eaeb69 [======================================] 2.0KiB / 2.0KiB
Copying blob e9b0018bd17e [======================================] 5.0KiB / 5.0KiB
Copying config 24276f54ce [======================================] 1.4KiB / 1.4KiB
Writing manifest to image destination
Storing signatures
--> 174d00599bba8a22fcb16a29de3dc6d125dec65f6b77d0748a6e9b7ea264a27d
```

**Comment by TomSweeneyRedHat** [comment]
Test issues were apparently self-induced, it's working fine for me too @QiWang19 .  Thanks for double checking.
LGTM

**Comment by TomSweeneyRedHat** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**